datashape/util/testing.py

```diff
@@ -0,0 +1,240 @@
+    String,
+    Time,
+    TimeDelta,
+    Tuple,
+    Units,
+)
+from ..dispatch import dispatch
+
+
+def _fmt_path(path):
+    """Format the path for final display.
+
+    Parameters
+    ----------
+    path : iterable of str
+        The path to the values that are not equal.
+
+    Returns
+    -------
+    fmtd : str
+        The formatted path to put into the error message.
+    """
+    if not path:
+        return ''
+    return 'path: _' + ''.join(path)
+
+
+@dispatch(DataShape, DataShape)
+def assert_dshape_equal(a, b, check_dim=True, path=None, **kwargs):
+    ashape = a.shape
+    bshape = b.shape
+
+    if path is None:
+        path = ()
+
+    if check_dim:
+        for n, (adim, bdim) in enumerate(zip(ashape, bshape)):
+            if adim != bdim:
+                path += '.shape[%d]' % n,
+                raise AssertionError(
+                    'dimensions do not match: %s != %s%s\n%s' % (
+                        adim,
+                        bdim,
+                        ('\n%s != %s' % (
+                            ' * '.join(map(str, ashape)),
+                            ' * '.join(map(str, bshape)),
+                        )) if len(a.shape) > 1 else '',
+                        _fmt_path(path),
+                    ),
+                )
+
+    path += '.measure',
+    assert_dshape_equal(
+        a.measure,
+        b.measure,
+        check_dim=check_dim,
+        path=path,
+        **kwargs
+    )
+
+
+class Slotted(with_metaclass(ABCMeta)):
+    @classmethod
+    def __subclasshook__(cls, subcls):
+        return hasattr(subcls, '__slots__')
+
+
+@assert_dshape_equal.register(Slotted, Slotted)
+def _check_slots(a, b, path=None, **kwargs):
+    """Genric checker that iterates over the ``__slots__`` and asserts they
+    are equal. This is a non-recursive function.
+
+    Parameters
+    ----------
+    a, b : Slotted
+        The shapes to check.
+    path : iteratable of str, optional
+        The path to the current ``a`` and ``b`` values.
+
+    Raises
+    ------
+    AssertionError
+        When the slots of ``a`` and ``b`` are not equal.
+    """
+    if type(a) != type(b):
+        return _base_case(a, b, path=path, **kwargs)
+
+    if a.__slots__ != b.__slots__:
+        raise AssertionError(
+            'slots mismatch: %r != %r\n%s' % (
+                a.__slots__, b.__slots__, _fmt_path(path),
+            ),
+        )
+    if path is None:
+        path = ()
+    for slot in a.__slots__:
+        if getattr(a, slot) != getattr(b, slot):
+            path += '.' + slot,
+            raise AssertionError(
+                "%s %ss do not match: %r != %r\n%s" % (
+                    type(a).__name__.lower(),
+                    slot,
+                    getattr(a, slot),
+                    getattr(b, slot),
+                    _fmt_path(path),
+                ),
+            )
+
+
+@assert_dshape_equal.register(object, object)
+def _base_case(a, b, path=None, **kwargs):
+    if a != b:
```

Comment:
Yup, I changed the rest to use this form, it is less verbose and better shows the intent.